src/db/file.rs

```diff
@@ -201,6 +210,18 @@ pub fn add_path_into_database<P: AsRef<Path>>(conn: &Connection,
                 ..Default::default()
             }).map_err(move |e| {
                 log::error!("failed to upload to {}: {:?}", bucket_path, e);
+
+                {
+                    // Get a new client, in case the old one's connection is stale.
+                    // AWS will kill our connection if it's alive for too long; this avoids
+                    // that preventing us from building the crate entirely.
+                    let mut c = retry_lock.write().unwrap();
```

Comment:
If that happens, how do we know which files to re-upload? Aren't the futures consumed in the call to `join_all`?

Reply:
We can reupload all the files; we might duplicate work but it shouldn't really matter (i.e., the files are the same).

Reply:
There are two places where we might want to re-create the client (two different `rt.block_on(::futures::future::join_all(futures))` calls):

* `if futures.len() > MAX_CONCURRENT_UPLOADS` - this is actually inside the `for file_path in ...` loop. after blocking on the initial futures, it clears the futures vec and continues with the next file (are you calling this the "completion of the inner loop"?)
* immediately following the `for file_path in ...` loop

were you talking about re-creating the client in both cases?

Reply:
when `MAX_CONCURRENT_UPLOADS` is reached, there's a partially-completed iterator that represents the files that we tried to upload (of which one or more failed). it's not immediately obvious to me how we would retry those uploads...

Reply:
Also, we currently have a retry for each file. It's unclear how we'd perform a reupload of all the files. It seems like we basically need to start over iterating through the file paths. Am I missing something?

Reply:
This looks like it doesn't match the code precisely -- but this is what I was envisioning in an "abstract" sense:

 * we have a list of all the file paths we need to upload (`to_upload`)
 * we remove MAX_CONCURRENT_UPLOADS elements off of this list (`currently_uploading`)
 * create a fresh client here (every time)
 * loop through these, creating a future for each upload
 * block on the completion of all these futures
   * if any fail, re-create the futures from the list `currently_uploading` (this may need an extra copy or so but that's fine) -- from the "create a fresh client" step basically
 * repeat the last several steps until we're done (i.e., `to_upload` is empty)

Reply:
That makes sense, I gave it a shot. It was a bit of a rewrite so hopefully it's correct :smile: